consensus/ising/synccache.go

```diff
@@ -1,103 +1,212 @@
+		maxHeight:  0,
+		blockCache: make(map[uint32]*BlockWithVotes),
+		voteCache:  make(map[uint32]map[uint64]Uint256),
+		timeLock:   NewTimeLock(),
 	}
 }
 
-// BlockInSyncCache returns whether the block exists in cache pool.
-func (sc *SyncCache) BlockInSyncCache(hash Uint256, height uint32) bool {
-	sc.RLock()
-	defer sc.RUnlock()
-
-	if blocks, ok := sc.cache[height]; ok {
-		for _, b := range blocks {
-			if hash.CompareTo(b.Hash()) == 0 {
-				return true
+// BlockInSyncCache returns block info and true is the block exists in cache.
+func (sc *SyncCache) BlockInSyncCache(hash Uint256, height uint32) (*BlockInfo, bool) {
+	if blockWithVotes, ok := sc.blockCache[height]; ok {
+		for _, b := range blockWithVotes.blockInfo {
+			if hash.CompareTo(b.block.Hash()) == 0 {
+				return b, true
 			}
 		}
 	}
 
-	return false
+	return nil, false
 }
 
 // CachedBlockHeight returns cached block height.
 func (sc *SyncCache) CachedBlockHeight() int {
 	sc.RLock()
 	defer sc.RUnlock()
 
-	return len(sc.cache)
+	return len(sc.blockCache)
 }
 
-// GetBlockFromSyncCache returns cached block which height is minimum.
-func (sc *SyncCache) GetBlockFromSyncCache() *ledger.Block {
+// GetBlockFromSyncCache returns cached block by height.
+func (sc *SyncCache) GetBlockFromSyncCache(height uint32) (*ledger.Block, error) {
+	err := sc.timeLock.WaitForTimeout(height)
+	if err != nil {
+		return nil, err
+	}
+
 	sc.RLock()
 	defer sc.RUnlock()
+	if _, ok := sc.blockCache[height]; !ok {
+		return nil, fmt.Errorf("no block in cache for height: %d", height)
+	}
 
-	if blocks, ok := sc.cache[sc.minHeight]; !ok {
-		return nil
-	} else {
-		if len(blocks) == 0 {
-			return nil
-		}
-		// TODO: return block which got enough votes
-		return blocks[0]
+	// check if there's a block got enough votes
+	if sc.blockCache[height].bestBlock == nil {
+		return nil, fmt.Errorf("ambiguous block for height: %d", height)
 	}
+
+	return sc.blockCache[sc.minHeight].bestBlock.block, nil
 }
 
-// AddBlockToSyncCache add received block to cache. Returns nil if block already existed.
-func (sc *SyncCache) AddBlockToSyncCache(block *ledger.Block) error {
+// AddBlockToSyncCache caches received block and the voter count when receive block.
+// Returns nil if block already existed.
+func (sc *SyncCache) AddBlockToSyncCache(block *ledger.Block, totalVoterCount int) error {
+	sc.Lock()
+	defer sc.Unlock()
+
 	hash := block.Hash()
 	blockHeight := block.Header.Height
-	if sc.BlockInSyncCache(hash, blockHeight) {
+	if _, exist := sc.BlockInSyncCache(hash, blockHeight); exist {
 		return nil
 	}
 
-	sc.Lock()
-	defer sc.Unlock()
-	if len(sc.cache) == 0 {
+	if len(sc.blockCache) == 0 {
+		// cached block height [min height, max height)
 		sc.minHeight = blockHeight
-		sc.maxHeight = blockHeight
+		sc.maxHeight = blockHeight + 1
 	} else {
-		if blockHeight != sc.maxHeight+1 {
+		if blockHeight != sc.maxHeight {
 			return errors.New("adding block which height is invalid")
 		}
 		sc.maxHeight++
```

Comment:
What if node receives multiple blocks of the same height?

Reply:
All received blocks with same height will be cached. Only the block which has >50% votes will be persisted.

Reply:
But is "sc.maxHeight++" called every time when the node receives a block and try to add it to sync cache?

Reply:
Good catch, it's a bug, let me fix it.